### PR TITLE
tiny gradcheck

### PR DESCRIPTION
**Pull request by 0xNaN**
Hi there, my attempt for `gradcheck`. It is really simple: it computes the approx. of the Jacobian with Finite-Differences and checks if it is close to the one computed using `.backward()`. (It should also work with `Conv2d` when it will be ready...not sure though). 

Confidence: 70% :smile: it was funny.

### CODE REVIEW

File: test/test.py [hunk 1]
```diff
@@ -32,6 +33,47 @@ def test_pytorch():
...
+        super(TorchModel, self).__init__()
+        self.l1 = torch.nn.Linear(*weights_init.shape, bias = False)
+        self.l1.weight = torch.nn.Parameter(torch.tensor(weights_init.T, requires_grad = True))
+      def forward(self, x):
+        return torch.nn.functional.log_softmax(self.l1(x).relu(), dim=1)
+
+    layer_weights = np.random.RandomState(1337).random((10, 5))
+    input_data = np.random.RandomState(7331).random((1, 10)) - 0.5
+
+    torch_input = torch.tensor(input_data, requires_grad = True)
```
Comment (geohot) [inline test/test.py:54]: My understanding is that gradcheck shouldn't have to compare to torch. Can it just compare to tinygrad in this test? (perhaps a different test to compare to torch if you want one?)

File: tinygrad/gradcheck.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      input : An input
+      eps : Perturbation step
+
+  returns:
+
+      NJ : an approx. of the Jacobian
+
+  [1]: https://timvieira.github.io/blog/post/2017/04/21/how-to-test-gradient-implementations/
+  """
+  output = model.forward(input)
```
Comment (geohot) [inline tinygrad/gradcheck.py:24]: Why is a model being passed in here? gradcheck shouldn't operate at this level of abstraction, if you need to run it multiple times, perhaps pass in a function instead?

File: tinygrad/gradcheck.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  ji = input.data.reshape(-1).shape[-1]
+  jo = output.data.reshape(-1).shape[-1]
+  NJ = np.zeros((ji, jo))
+
+  for i in range(ji):
+    for o in range(jo):
+
+      eps_pertub = mask_like(input.data, i, mask_value = eps)
```
Comment (geohot) [inline tinygrad/gradcheck.py:33]: perturb

**Comment by geohot** [comment]
I think we can get this whole PR below 50 lines :)

File: tinygrad/gradcheck.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    # tinygrad doesn't support slicing, tiny-hack to select
+    # the needed scalar an backpropagate only through it
+    o_scalar = Tensor(mask_like(output.data, o, 1.)).mul(output).sum()
+    o_scalar.backward()
+
+    for i, grad in enumerate(input.grad.reshape(-1)):
+      J[i][o] = grad
+
+  test_passed = np.allclose(J, NJ, atol=atol, rtol=rtol)
+  return test_passed, J, NJ
```
Comment (geohot) [inline tinygrad/gradcheck.py:75]: I'd factor out jacobian as a function, make this function small, then just return a bool to match torch

**Comment by geohot** [comment]
Ahh, sorry. I merged the test PR. Rebase? It looks good

**Comment by 0xNaN** [comment]
Done. Now we have:

1. `jacobian()` (analytical) and `numerical_jacobian()` as extracted functions
2. better and refactored tests
3. interface aligned with Pytorch. One difference here is that pytorch can compute Jacobian and gradcheck of functions of _multiple tensors_  (i.e: `func(ta, tb, ...)`) while this version does not (for this reason the parameter is named _input_ rather than _inputs_). Seems an easy extension: just compute the current Jacobian (e.g: analytical) of the output of `func(ta, tb, ...)` w.r.t. each tensor individually while collecting the Jacobians in a list. I will do it soon, pytorch example as reference:

```python
>>> t1 = torch.tensor([1., 1., 1.], requires_grad = True)
>>> t2 = torch.tensor([2.], requires_grad = True)
>>> func = lambda x, y: (x*y)
>>> torch.autograd.functional.jacobian(func, (t1, t2))
(tensor([[2., 0., 0.],
        [0., 2., 0.],
        [0., 0., 2.]]), tensor([[1.],
        [1.],
        [1.]]))
>>> torch.autograd.functional.jacobian(lambda k: func(k, t2), t1)
tensor([[2., 0., 0.],
        [0., 2., 0.],
        [0., 0., 2.]])
>>> torch.autograd.functional.jacobian(lambda k: func(t1, k), t2)
tensor([[1.],
        [1.],
        [1.]])
```

PS: comma will win ;)

**Pull request merged**